arangod/Graph/KPathFinder.cpp

```diff
@@ -0,0 +1,339 @@
+                   std::make_move_iterator(_shell.end()));
+  _shell.clear();
+  _depth++;
+}
+
+auto KPathFinder::Ball::noPathLeft() const -> bool {
+  return doneWithDepth() && _shell.empty();
+}
+
+auto KPathFinder::Ball::doneWithDepth() const -> bool {
+  return _searchIndex >= _interior.size();
+}
+
+auto KPathFinder::Ball::getDepth() const -> size_t { return _depth; }
+
+auto KPathFinder::Ball::shellSize() const -> size_t { return _shell.size(); }
+
+auto KPathFinder::Ball::buildPath(VertexIdentifier const& vertexInShell,
+                                  PathResult& path) -> void {
+  VertexIdentifier const* myVertex = &vertexInShell;
+  if (_direction == Direction::FORWARD) {
+    while (myVertex->predecessor != 0 || myVertex->id != _center) {
+      path.prependVertex(myVertex->id);
+      path.prependEdge(myVertex->edge);
+      TRI_ASSERT(_interior.size() > myVertex->predecessor);
+      myVertex = &_interior[myVertex->predecessor];
+    }
+    path.prependVertex(_center);
+  } else {
+    // For backward we just need to attach ourself
+    // So everything until here should be done.
+    // TRI_ASSERT(!path.empty());
+    if (myVertex->predecessor == 0 && myVertex->id == _center) {
+      // already reached the center
+      return;
+    }
+    TRI_ASSERT(_interior.size() > myVertex->predecessor);
+    path.appendEdge(myVertex->edge);
+    myVertex = &_interior[myVertex->predecessor];
+    while (myVertex->predecessor != 0 || myVertex->id != _center) {
+      path.appendVertex(myVertex->id);
+      path.appendEdge(myVertex->edge);
+      TRI_ASSERT(_interior.size() > myVertex->predecessor);
+      myVertex = &_interior[myVertex->predecessor];
+    }
+    path.appendVertex(_center);
+  }
+}
+
+auto KPathFinder::Ball::matchResultsInShell(VertexIdentifier const& match,
+                                            ResultList& results) const -> void {
+  auto [first, last] = _shell.equal_range(match);
+  if (_direction == FORWARD) {
+    while (first != last) {
+      results.push_back(std::make_pair(*first, match));
+      first++;
+    }
+  } else {
+    while (first != last) {
+      results.push_back(std::make_pair(match, *first));
+      first++;
+    }
+  }
+}
+
+auto KPathFinder::Ball::computeNeighbourhoodOfNextVertex(Ball const& other,
+                                                         ResultList& results) -> void {
+  TRI_ASSERT(!doneWithDepth());
+  auto const& vertex = _interior[_searchIndex].id;
+  _cursor->rearm(vertex.stringRef(), 0);
+  _cursor->readAll([&](EdgeDocumentToken&& eid, VPackSlice edge, size_t /*cursorIdx*/) -> void {
+    VertexRef id = _cache->persistString(([&]() -> auto {
+      if (edge.isString()) {
+        return VertexRef(edge);
+      } else {
+        VertexRef other(transaction::helpers::extractFromFromDocument(edge));
+        if (other == vertex) {
+          other = VertexRef(transaction::helpers::extractToFromDocument(edge));
+        }
+        return other;
```

Comment:
Still wondering if the following code is equivalent to the existing code...
```
diff --git a/arangod/Graph/KPathFinder.cpp b/arangod/Graph/KPathFinder.cpp
index 5ca733ab76..a893359cf0 100644
--- a/arangod/Graph/KPathFinder.cpp
+++ b/arangod/Graph/KPathFinder.cpp
@@ -188,11 +188,17 @@ auto KPathFinder::Ball::computeNeighbourhoodOfNextVertex(Ball const& other,
       if (edge.isString()) {
         return VertexRef(edge);
       } else {
-        VertexRef other(transaction::helpers::extractFromFromDocument(edge));
+        if (other._direction == Direction::FORWARD) {
+          return VertexRef(transaction::helpers::extractFromFromDocument(edge));
+        } else {
+          return VertexRef(transaction::helpers::extractToFromDocument(edge));
+        }
+        /*
         if (other == vertex) {
           other = VertexRef(transaction::helpers::extractToFromDocument(edge));
         }
         return other;
+        */
       }
     })()); 
```

Reply:
I think it is not.
The "FORWARD" here is not defining that we go from _from to _to.
It is defining that we go in the direction the user has written.
e.g. OUTBOUND (which is _from to _to) but if the user has written INBOUND the above suggestion would return the wrong vertex.

We plan on improving this in a different PR once and for all.

Reply:
Vale